src/coreclr/jit/lowerxarch.cpp

```diff
@@ -3693,6 +3727,73 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
         LowerNode(cast);
     }
 }
+
+//----------------------------------------------------------------------------------------------
+// Lowering::LowerHWIntrinsicToScalar: Lowers a tree AND(X, ADD(X, -1) to HWIntrinsic::ResetLowestSetBit
+//     Returns the recplacement node if one is created else nullptr indicating no replacement
+//
+// Parameters
+//     tree   -   GentreePtr of binOp
+//
+GenTree* Lowering::LowerAndOpToResetLowestSetBit(GenTree* node)
+{
+    GenTree* op1 = node->gtGetOp1();
+    GenTree* op2 = node->gtGetOp2();
+    if (op1->OperIs(GT_LCL_VAR) && op2->OperIs(GT_ADD) && !comp->lvaGetDesc(op1->AsLclVar())->IsAddressExposed())
+    {
+        GenTree* addOp1 = op2->gtGetOp1();
+        GenTree* addOp2 = op2->gtGetOp2();
+        if (addOp2->IsIntegralConst(-1) && addOp1->OperIs(GT_LCL_VAR) &&
+            addOp1->AsLclVar()->GetLclNum() == op1->AsLclVar()->GetLclNum() &&
+            ((op1->TypeGet() == TYP_LONG && comp->compOpportunisticallyDependsOn(InstructionSet_BMI1_X64)) ||
+             comp->compOpportunisticallyDependsOn(InstructionSet_BMI1)) // if op1->TypeGet()!=TYP_LONG then it must be
+                                                                        // TYP_INT
+            )
+        {
+            LIR::Use use;
+            if (BlockRange().TryGetUse(node, &use))
+            {
+                // If we allow the use of BSLR where the parent is NE or EQ it will generate "blsr, test"
+                // if we prevent the use of BSLR here then the OptimizeConstCompare lowering may generate
+                //   TEST_NE or TEST_EQ leading to a cheaper "lea, test"
+                GenTree* user = use.User();
+                if (user != nullptr && !user->OperIs(GT_EQ, GT_NE))
+                {
+                    GenTreeHWIntrinsic* replacementNode =
+                        comp->gtNewScalarHWIntrinsicNode(op1->TypeGet(), op1,
+                                                         op1->TypeGet() == TYP_INT
+                                                             ? NamedIntrinsic::NI_BMI1_ResetLowestSetBit
+                                                             : NamedIntrinsic::NI_BMI1_X64_ResetLowestSetBit);
```

Comment:
This should use `node`s types, both for the new node and the intrinsics. The reason is that for `AND` we're guaranteed to only ever see `TYP_INT/TYP_LONG` (with one exception that is not relevant here), while `op1` can be of a small type (`TYP_UBYTE` and friends).

Nit: `TypeGet()` -> `TypeIs`

Reply:
All done.